client/DESIGN.md

```diff
@@ -0,0 +1,292 @@
+
+We provide an overview of the Cache class below, including its primary internal data structure, its public methods, and the form of its HTTP requests and responses.
+
+- Parameters: 
+    1. server: string (required)
+    2. port: string (port=80)
+    3. max\_age: int (max\_age=None)
+    4. no\_cache: bool (no\_cache=False)
+- Public Methods: 
+    1. read(table: string, key: string, read\_timestamp=currentTime, max\_age=None, no\_cache=False): value (JSON)
+    2. write(table: string, op\_list: list): status (string), throws StaleException
+- Internal Objects: 
+    1. HTTP library for establishing/managing connection to server (in Python, urllib3)
+    2. sorted map for maintaining cache (in Python, [skiplist](https://github.com/sepeth/python-skiplist))
+- Exceptions: 
+    1. StaleException
+    2. StaleException(Value_time: int)
+
+## Cache Structure Description
+
+The cache internally maintains (1) a connection to the DB server and (2) a sorted key-value map of cache entries.  
+
+### Database Connection
+
+The cache should utilize an HTTP Request/Response library (e.g., in Python, urllib3) to establish a connection with the DB server and read and write values through its course of use.  We describe in the HTTP Message Formatting section the expected format for DB requests and responses.
+
+[Here are examples](https://urllib3.readthedocs.org/en/latest/) of using the urllib3 library in Python to open and use a connection.
+
+### Sorted Key-Value Map of Cache Entries
+
+The cache is a ("table:key" map) of (sorted value maps), which we explain below.
+
+The keys of the "table:key" map are composed of the concatenation of the table identifier string and the key string; each "table:key" maps to a (sorted value map).  (Note that the cache may store entries from multiple tables.)  
+
+The cache entries themselves are stored in (sorted value maps).  The keys of the (sorted value maps) are derived from the read timestamps of each entry.  Since we want the entries sorted from most-recent to oldest read timestamps, the actual keys are longs with the value 
+
+key = (LONG\_MAX - read\_timestamp)
+
+For example, in Python, the outer "table:key" map can be implemented as a standard Python dictionary, and we just use string concatenation (a+b) to form the keys.  For the inner sorted value maps, we can use a SortedDict implementation based on [skiplists](https://github.com/sepeth/python-skiplist).
+
+* **Question (1):** We said Tuesday we just need a single-level map, but why don't we need at least the two levels I describe here?
+
+## Cache Public Method Descriptions
+
+### read
+    read(
+        table: string, 
+        key: string, 
+        read_timestamp=currentTime, 
+        max_age=None, 
+        no_cache=False): value (JSON)  
+
+The read method sends a read request to the DB for the given table and key, along with the specified cache parameters, if any.  It returns the value object if the DB provides a value; otherwise, it returns None.
```

Comment:
> The read method sends a read request to the DB

The read first checks the cache. It sends a request to the DB if the entry is missing, or if its been too long since we read that entry.